### DISCUSSION TITLE
Using with Laravel Spark - "Route login not defined" Error

### DISCUSSION THREAD

**Discussion by twd3** [jacksleight/statamic-members#15] [General]
Your addon's are the best! This is so easy to get going with Laravel Spark! There is only one thing I'm having an issue with.

If I'm NOT logged in, and I try to visit /billing I get a "Route login not defined" error.

I found John's awesome article about integrating Spark with Statamic and he ran in to this same issue [post here](https://stillat.com/blog/2021/02/20/integrating-laravel-spark-and-statamic-3).

None of the things he did to fix this works for me so it has to be something with Memberbox. I don't know why though because both Statamic, and Memberbox just have a login route of /login.

One last thing is John also talked about adding something to the web.php route file to temp fix this and that did not work for me either. [discord post here](https://discord.com/channels/489818810157891584/489818810157891586/812735607322836992)

Any insight on this issue would be greatly appreciated!

**Comment by jacksleight** [comment]
Hey Tom. I'm a little unsure about this one, I'm not familiar with Spark. Memberbox doesn't create a route named `login`, Memberbox's route is called `statamic-memberbox.login`. Are you saying that your existing app has a route called `login`, but as soon as you install Memberbox you then start getting that error?

**Reply by twd3** [reply]
Well, existing app is Statamic, and they have the logged in protect thing:
```
        'logged_in' => [
            'driver' => 'auth',
            'login_url' => '/login',
            'append_redirect' => true,
        ],
```
So there's that login route (/login) right? Then in Memberbox config there is the login route, which I have set to /login as well.

So there are "login" routes and when you hit something protected by a login like /billing url for Spark and you are NOT logged in it should take you to a login page. So I don't know why Spark thinks there is no login route.

![Picture1](https://user-images.githubusercontent.com/16004958/203620453-9fb585f2-e35b-4e4c-a0fa-dea15b9d856f.png)

**Reply by jacksleight** [reply]
> Then in Memberbox config there is the login route

There is, but internally the name of that route is namespaced as `statamic-memberbox.login`. If Spark is looking for a route called `login` it's not going to find the Memberbox one. Is there a way to tell Spark the name of your login route? If so set it to `statamic-memberbox.login`.

If not I'd suggest switching off the automatic Memberbox route (by commenting out the `'login' => '/login'` line in the config), and then defining it yourself in `routes/web.php`:

```php
Route::statamic('/login', 'statamic-memberbox::web.login')->name('login');
```

**Reply by twd3** [reply]
Okay, tried that and got "Route [statamic-memberbox.login] not defined."

**Reply by jacksleight** [reply]
Ah OK, I assume you're also using the `{{ mb:user:login_url }}` tag? That depends on the Memberbox route.

If you replace that with `{{ route:login }}` it'll use your custom login route instead.

I'll have a think about whether Memberbox could make this simpler. The routes are deliberately namespaced so they don't conflict with other application routes, which is generally a good idea. But if you have other packages (like Spark) that assume your routes are named a particular way this isn't ideal. Perhaps I could make it possible to opt-out of the namespacing.

**Reply by twd3** [reply]
I was using that in my nav but this doesn't happen when I click a nav link. It happens when I visit /billing when not logged in. So I don't think using ```{{ route:login }}``` would work. It's like Spark doesn't see the memberbox OR the Statamic login so it can't re-direct.

**Reply by jacksleight** [reply]
So there's two issues happening here, making both of those changes should get things up and running.

**Route login not defined**
This is because Spark is looking for `login`, but Memberbox's route is called `statamic-memberbox.login`. Commenting out the line in the config and then defining the route in your routes file fixes this.

However, doing that creates this error:

**Route statamic-memberbox.login not defined**
This is because commenting out that line removed the MB route, but the MB tag (that you're using in your nav) requires it. Switching out that tag for the route tag will fix this.

Then it should all be good.

Like I say, I'll look into a nicer way to handle this in MB. If you could just call the MB route `login` it would all just work without any other changes.

**Comment by twd3** [comment]
Oooooo, I tried this and it DID forward me to the CP login!

```
Route::get('/login', function () {
    return redirect(route('statamic.cp.login'));
})->name('login');
```
So what do you think would be the fix?

**Comment by twd3** [comment]
But this doesn't work:
```
Route::get('/login', function () {
    return redirect(route('statamic-memberbox::web.login'));
})->name('login');
```
Or this:
```
Route::get('/login', function () {
    return redirect(route('statamic.login'));
})->name('login');
```
^ This is weird because why would statamic.cp.login work but not statamic.login?

**Comment by twd3** [comment]
Ha, got it! Just kept trying stuff lol. So this works GREAT with Spark. Now that I know FOR sure what to do I can have a paid membership site's backend done in under an hour! Fix was in /app/Http/Middleware/Authenticate.php. Had to change it to:
```
<?php

namespace App\Http\Middleware;

use Illuminate\Auth\Middleware\Authenticate as Middleware;

class Authenticate extends Middleware
{
    /**
     * Get the path the user should be redirected to when they are not authenticated.
     *
     * @param  \Illuminate\Http\Request  $request
     * @return string|null
     */
    protected function redirectTo($request)
    {
        if (! $request->expectsJson()) {
            return route('statamic-memberbox.login');
        }
    }
}
```
Thank you for your help with this and ALL your AWESOME addons!

**Reply by jacksleight** [reply]
Ah ha, perfect! I didn't know you could change it there. Thanks for figuring it out, I've added that to the docs as well: https://jacksleight.dev/docs/memberbox/configuration#using-laravel-middleware-to-redirect-unauthenticated-users

**Comment by ItsJustPW** [comment]
Just writing a little comment on here because this is exactly what I'm looking to do (paid member site) but I'm a very new dev so at least I know it can be done with Statamic, MemberBox and Spark!  If anyone has a useful guide on implementation... I'd be happy to buy you a coffee! : ) 

Thanks again... great work guys!
